src/Functions/mortonCoding.cpp

```diff
@@ -0,0 +1,326 @@
+constexpr auto MortonND_5D_Dec = mortonnd::MortonNDLutDecoder<5, 12, 8>();
+constexpr auto MortonND_6D_Dec = mortonnd::MortonNDLutDecoder<6, 10, 8>();
+constexpr auto MortonND_7D_Dec = mortonnd::MortonNDLutDecoder<7, 8, 8>();
+constexpr auto MortonND_8D_Dec = mortonnd::MortonNDLutDecoder<8, 8, 8>();
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int ILLEGAL_COLUMN;
+    extern const int ARGUMENT_OUT_OF_BOUND;
+}
+
+namespace
+{
+class FunctionMortonEncode : public IFunction
+{
+public:
+    static constexpr auto name = "mortonEncode";
+    static FunctionPtr create(ContextPtr)
+    {
+        return std::make_shared<FunctionMortonEncode>();
+    }
+
+    String getName() const override
+    {
+        return name;
+    }
+
+    bool isVariadic() const override
+    {
+        return true;
+    }
+
+    size_t getNumberOfArguments() const override
+    {
+        return 0;
+    }
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DB::DataTypes & arguments) const override
+    {
+        for (const auto& arg: arguments)
+        {
+            if (!WhichDataType(arg).isNativeUInt())
+                throw Exception("Illegal type " + arg->getName() + " of argument of function " + getName() + ", expected UInt[8-64]",
+                                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        }
+        return std::make_shared<DataTypeUInt64>();
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+
+#define EXTRACT_VECTOR(INDEX) \
+    const ColumnPtr & col##INDEX = non_const_arguments[(INDEX)].column;
+
+#define ENCODE(ND, ...) \
+    if (arg_size == (ND)) \
+    { \
+        for (size_t i = 0; i < input_rows_count; i++) \
+        {               \
+            vec_res[i] = MortonND_##ND##D_Enc.Encode(__VA_ARGS__); \
+        } \
+        return col_res; \
+    }
+
+        size_t arg_size = arguments.size();
+
+        auto non_const_arguments = arguments;
+        for (auto & argument : non_const_arguments)
+            argument.column = argument.column->convertToFullColumnIfConst();
+
+        auto col_res = ColumnUInt64::create();
+        ColumnUInt64::Container & vec_res = col_res->getData();
+        vec_res.resize(input_rows_count);
+
+        EXTRACT_VECTOR(0)
+        if (arg_size == 1)
+        {
+            // just convert to UInt64 column
+            for (size_t i = 0; i < input_rows_count; i++)
+            {
+                vec_res[i] = col0->getUInt(i);
+            }
+            return col_res;
+        }
+
+        EXTRACT_VECTOR(1)
+        ENCODE(2, col0->getUInt(i), col1->getUInt(i))
```

Comment:
Do we align arguments by the most significant bit?
So, if we are mixing UInt8 and UInt32, it will not start with 24 nonexistent bits of the UInt8 argument?

Reply:
I'm not sure if the library is stable in regards to variable length arguments. Will write tests and see. At least I know that its AVX2 and non-AVX code is stable. Will check all the interesting cases.

Reply:
If you are obtaining UInt64 values with IColumn::getUInt, then how the library will know about the original type?

Reply:
It does LUT by 8 bytes.
Seems fine:
```
SELECT mortonDecode(4, mortonEncode(CAST('255', 'UInt8'), CAST('256', 'UInt32'), CAST('15', 'UInt8'), CAST('15', 'UInt32')))

Query id: 416b0c9c-aed7-4cb3-af38-876d919841b2

┌─mortonDecode(4, mortonEncode(CAST('255', 'UInt8'), CAST('256', 'UInt32'), CAST('15', 'UInt8'), CAST('15', 'UInt32')))─┐
│ (255,256,15,15)                                                                                                       │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 0.001 sec. 
```
Can you give an example where it should decode wrong?